File: cplusplus/040_Combination_Sum_II.cpp
/*
For each number, we have two options, choose or not choose.
This can form a decision tree.
For each tree node, if its sum is equal to target, we can add 
it to final result.
f[1,1,1,2,2] = (1 , f[1,1,2,2]),
               (1,f[1,2,2]),
               (1,f[2,2]),
               (2,f[2]),
               (2,f[])
Here, the second and thrid item are the subset of f[1,1,2,2].
The final item is the subset fourth item.
*/

class Solution {
public:
  vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
    vector<vector<int>> res;
    vector<int> temp;
    std::sort(candidates.begin(),candidates.end());
    dfs(res,temp,candidates,target,0);
    return res;
  }
private:
  void dfs(vector<vector<int>> & res,
           vector<int> & temp,
           const vector<int> & candidates,
           int target,
           int start){
    if (target==0) {
      res.push_back(temp);
    } else {
      for (int i=start;i<candidates.size();++i) {
        if (target<candidates[i]) break;
        if (i>start && candidates[i]==candidates[i-1]) continue;
        temp.push_back(candidates[i]);
        dfs(res,temp,candidates,target-candidates[i],i+1);
        temp.pop_back();
      }
    }
  }
};

File: java/040_Combination_Sum_II.java
// Solution one, backtracking
public class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> reval = new ArrayList<List<Integer>>();
        Arrays.sort(candidates);
        checkWithBacktracking(reval, new ArrayList<Integer>(), candidates, 0, target);
        return reval;
    }

    private void checkWithBacktracking(List<List<Integer>> reval,
                                       ArrayList<Integer> temp,
                                       int[] candidates,
                                       int start,
                                       int target) {
        if (target == 0 && !temp.isEmpty()) {
            reval.add(new ArrayList<Integer>(temp));
        } else {
            for (int i = start; i < candidates.length && target >= candidates[i]; ++i) {
                if (i > start && candidates[i] == candidates[i-1]) continue; // Avoid duplicates
                temp.add(candidates[i]);
                checkWithBacktracking(reval, temp, candidates, i + 1, target - candidates[i]);
                temp.remove(temp.size() - 1);
            }
        }
    }
}

